### PR TITLE
Optimize threading in AbstractSearchAsyncAction

### PR DESCRIPTION
**Pull request by original-brownbear**
Forking when an action completes on the current thread is needlessly heavyhanded in preventing stack-overflows. Also, we don't need locking/synchronization to deal with a worker-count + queue length problem. Both of these allow for non-trivial optimization even in the current execution model, also this change helps with moving to a more efficient execution model by saving needless forking to the search pool in particular (think about the latency impact of having a full search queue on the local node at the same time, blocking fanning out of requests to shards on other nodes).
-> refactored the code to never fork but instead avoid stack-depth issues through use of a `SubscribableListener`
-> replaced our home brew queue and semaphore combination with JDK primitives which saves blocking synchronization on task start and completion.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java [hunk 1]
```diff
@@ -824,61 +792,63 @@ protected final ShardSearchRequest buildShardSearchRequest(SearchShardIterator s
...
             assert permits > 0 : "not enough permits: " + permits;
-            this.permits = permits;
+            semaphore = new Semaphore(permits);
         }
 
-        Runnable finishAndGetNext() {
-            synchronized (this) {
-                permitsTaken--;
-                assert permitsTaken >= 0 : "illegal taken permits: " + permitsTaken;
+        void submit(Consumer<Releasable> task) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:804]: `Consumer<Releasable>` is a slightly lazy choice here admittedly, but I found it easier to read to have a `Releasable` in the actual execution than passing the listener and only calling `onResponse` on it no matter what.
Also, `Releasable` has try-with-resources :)

**Comment by original-brownbear** [comment]
@elasticmachine update branch

**Review by javanna — COMMENTED** [review]
Can this change perhaps be split into prerequisite refactoring + pending executions adjustments? Also, could you clarify what this PR aims to change and how? How does removing the fork compare to the problem that forking addressed (I believe #27069 introduced it)?.

**Comment by original-brownbear** [comment]
I can try :) It's a little intertwined but I'll pull out what I can first without doing the actual execution mode change.

The heart of the fix is here: https://github.com/elastic/elasticsearch/pull/113230/files#diff-96633e6da7d4c7696cf8d66d892f86fab7e407f42b45bb9d1df9b6b92b1270d5R821 we identify the case where a task executes synchronously and loop on the queue of tasks if that's the case. So we never execute synchronously and recursively, it's either of the two, loop when sync, callback when async.

**Comment by original-brownbear** [comment]
Hmm actually sorry @javanna this is actually already my cleaned up version of this, my bad, my memory isn't so great these days :) I don't think I there's much in the way of preliminary work in here that I can reasonably pull out.

**Review by piergm — APPROVED** [review]
LGTM! 😄

**Comment by original-brownbear** [comment]
Thanks Matteo!

**Pull request merged**